### PR TITLE
[WIP] Add enterprise-client to Telegraf

### PR DESCRIPTION
**Pull request by timraymond**
This adds a new `services` subdirectory to Telegraf, where services that exist
alongside plugins can perform background tasks. Specifically, this is to
facilitate registration with Enterprise, and hosting the administration
interface that Enterprise uses to communicate with Telegraf. 

At the moment, registration and running the admin interface works. However, we
have placeholder values for ProductID and ClusterID, which are currently
required by Enterprise. We want these values to be stable, but randomly set if
not provided. Kapacitor handles this by writing both IDs to a `server-id` and
`cluster-id` file under `~/.kapacitor` (see:
https://github.com/influxdata/kapacitor/blob/stats-enterprise/cmd/kapacitord/run/server.go#L461-L485).
We could take the same approach here, but since I'm not sure what the best way
forward is, I figured I would open it up to discussion.

### CODE REVIEW

**Comment by sparrc** [comment]
you need to add github.com/influxdata/enterprise-client/v2 to Godeps

**Comment by sparrc** [comment]
@timraymond can you do a `gdm restore` before running `gdm save`? So that you don't change the hash on all of these unrelated dependencies

**Comment by sparrc** [comment]
Also why does the enterprise client add 7 new dependencies? What is it doing exactly?

File: agent/agent.go [hunk 1]
```diff
@@ -328,6 +330,9 @@ func (a *Agent) Run(shutdown chan struct{}) error {
 	}
 	ticker := time.NewTicker(a.Config.Agent.Interval.Duration)
 
+	ent := enterprise.NewEnterprise(a.Config.Agent.Enterprise, a.Config.Agent.Hostname, a.Version, shutdown)
+	go ent.Open()
```
Comment (sparrc) [inline agent/agent.go]: do we _always_ want to run this? What does it do?

**Comment by timraymond** [comment]
@sparrc Okay, that was weird because I ran `gdm restore` beforehand. It turned out that I didn't have remote tracking setup for the master branches of one of my repos, so I guess `gdm restore` stopped restoring when it hit that? Anyway, fixed now... I pushed up the reversions of those unnecessary updates, and I can squash it down prior to merging.

The seven new deps are mostly for the http handlers that we're setting up for the admin interface. We're using `echo` as a router, which pulls in `go-colorable` and `go-isatty` for logging, I believe. I'm not really sure what `gomman` is, but it's by the same guy that made `echo`, so I'm assuming `echo` uses it. `markbates/hmax` makes it a little easier to do HMAC signing of requests between the Enterprise client and Enterprise. `markbates/going` is... a little bit of a grab bag of packages that Mark has put together for writing Go web applications.

File: agent/agent.go (hunk 1 above)
Comment (timraymond) [inline agent/agent.go]: It checks the config to see if we should perform registration with enterprise and startup the admin interface. If there are no configured enterprise hosts, it just returns.

File: internal/config/config.go [hunk 2]
```diff
@@ -93,6 +94,8 @@ type AgentConfig struct {
 	// Quiet is the option for running in quiet mode
 	Quiet    bool
 	Hostname string
+
+	Enterprise enterprise.Config
```
Comment (sparrc) [inline internal/config/config.go]: So for this to be activated, the user would need an `enterprise` table underneath the `[agent]` table? What does that look like? Should we put a commented-out example in the sample config?

Comment (timraymond) [inline internal/config/config.go]: Might not be a bad idea—I'll add one. It basically looks like this:

```toml
[agent]
# snip...

  [agent.enterprise]
  client_port = 1234

    [[agent.enterprise.hosts]]
    url = "http://example.com:5678"
    primary = true # optional if you only have one host

    [[agent.enterprise.hosts]]
    url = "http://example.com:9876"
    primary = false
````

**Comment by timraymond** [comment]
@sparrc Any thoughts on where we should configure / store Product IDs and Cluster IDs? (see the description).

**Comment by sparrc** [comment]
`~/.telegraf` or something similar seems fine with me

**Comment by sparrc** [comment]
@timraymond The main concern I have with this is that it's not really actually a "client", it's a client and a _server_, if I understand correctly.

I think it's good to have an actual enterprise "client" in telegraf, but this goes quite a bit further and seems like it should be it's own service running side-by-side with telegraf.

**Comment by sparrc** [comment]
Also, I still don't feel like I understand what it's doing..? So it's starting up an http service? And then what? It's going to control service restarts? Configuration?

**Comment by sparrc** [comment]
@timraymond I believe we can close this for the time being?

**Comment by timraymond** [comment]
@sparrc yep, totally fine... we're moving in another direction with enterprise-client.

**Pull request closed without merging**